guide: begin How To section (#1705)
iterative/dvc.org: content/docs/user-guide/how-to/best-practices.md

# Best Practices for DVC Projects

DVC provides a systematic approach towards managing and collaborating on data
science projects. You can manage your projects with DVC more efficiently using

> the practices listed here:

Author: This file should be removed from here. That's the point of extracting this to #1748.

# Best Practices for DVC Projects

> Asking questions on data science collaboration to data scientists, engineers, or
> managers, we'll get a variety of answers. DVC provides a systematic approach

Author: No need to keep all the text fro mold versions if it no longer makes sense. This doc has nothing to do with Q&A now so:

```suggestion
DVC provides a systematic approach
```

towards managing and collaborating on data science projects. You can manage your
projects with DVC more efficiently using the practices listed here:

- Source code and data versioning

>   You can use DVC to avoid discrepancies between

Author: These bullets should probably be H2 headers instead, so each practice get's a right-hand nav entry. The doc is too long for it to be a single bullet list anyway.

> - Source code and data versioning
> 
>   You can use DVC to avoid discrepancies between
>   [revisions](https://git-scm.com/docs/revisions) of source code and versions of
>   data files, when the data doesn't fit into a traditional repository. DVC

Author: Nice. You changed the open question into an answer. Let's just be much more on point for how-tos. Some distracting notes should be omitted and you can link to other docs so please get very familiar with all of the existing docs! For example:

```suggestion
- Source code and data versioning

  You can use DVC to avoid discrepancies between
  [revisions](https://git-scm.com/docs/revisions) of source code and
  [versions](/doc/use-cases/versioning-data-and-model-files) of data files. DVC
```
> Needs formatting.

Author: BUT this point is still mostly explanation and not very precise solution. See https://documentation.divio.com/introduction/ again to have a clear understanding of the differences.

If you think Best Practices should be a regular guide and not a How-to, we can discuss that. Please don't just follow my instructions blindly, I'm expecting to be wrong sometimes and for you to drive this whole process!

Reviewer: Yea as I described in https://github.com/iterative/dvc.org/pull/1705#discussion_r479693987 its better to make it a regular guide. It can have some explanationa and example. if an example already exist that can be linked too.

Author: OK. I'm not sure but I'll let you lead this process. Curious to see what links you would add. As mentioned before, please make a new draft PR for that and leave your suggested changes there. Maybe I'll take it over then, as it's slightly advanced and not necessarily related to your GSoD project description (1.x updates and how-to).

  replaces all large data files, models, etc. with small
  [metafiles](doc/user-guide/dvc-files-and-directories) (tracked by Git). These
  files point to the original data, which you can access by checking out the
  required `revision`.

- Experiments

  You can make use of Git branches to document progress of training different
  types of models on your data files in the same project. Create a branch for

>   each of the model and then utilise DVC features while working on that branch.

Author: ```suggestion
  each of the models and then utilise DVC features while working on that branch.
```

>   each of the models and then utilise DVC features while working on that branch.

Author: This would need much more info. One experiment per branch is just one of the experimentation approaches. Please browse issues about Experiments both in this repo and in the core repo and/or ask us via Discord for more info.

Author: Also needs more on-point solutions and less explanation (if we're sticking with Best Practices as a How-to).

Reviewer: Removed this. This can actually be a single doc under How-to section. It can be implemented through branches, tag etc. Will take a lot space here with all explanation and steps.

Author: Hmmm... I think it's definitely one of the most important things to add in best practices. Although there is an issue about writing a separate doc about this indeed: #816 — but I think both can exist, and the section in Best Practices can just give an overview, and later link to the separate doc, when it exists... Please restore it for the new PR you'll be creating.

Author: Moved to https://github.com/iterative/dvc.org/pull/1748#pullrequestreview-483189340. Resolving here.

- Experiment time log

  [Hyperparameter](<https://en.wikipedia.org/wiki/Hyperparameter_(machine_learning)>)
  are defined using the the `--params` option of `dvc run` and the default

>   parameters file is `params.yaml`. You can commit different versions of

Author: Please introduce the problem first because it's not obvious from the title. In fact I don't really love this title... Should maybe use both "experiment" and "parameter" terms?

Author: "defined using the the `--params` option of `dvc run` and the default parameters file is `params.yaml`." - too much information. Try to simplify and mention only the absolutely necessary. Better to link to the params reference for the rest of the details.

>   parameters file is `params.yaml`. You can commit different versions of

Author: "defined using the the `--params` option of `dvc run` and the default parameters file is `params.yaml`." - too much information. Try to simplify and mention only the absolutely necessary. Better to link to the params reference for the rest of the details.

  `params.yaml` and then use `dvc metrics` or `dvc plots` to track which of your

>   changes contributed the most in improving target

Author: "defined using the the `--params` option of `dvc run` and the default parameters file is `params.yaml`." - too much information. Try to simplify and mention only the absolutely necessary. Better to link to the params reference for the rest of the details.

Same for "and then use `dvc metrics` or `dvc plots`" ... way too much information. A first-time user would have no idea how these things connect.

Author: Probably better to add code blocks. Remember, this is supposed to be a how-to...

Reviewer: Simplified this in new update with code blocks. haven't added outputs (only command)

Author: Command should be enough. It's not a tutorial, not even a how-to as we've just recently decided. Resolving for now.

  [metric](doc/command-reference/metrics). You can monitor the degree of each
  change.

- Navigating through experiments

  To recover a model from last week without wasting time required for the model
  to retrain, first you can checkout the required `revision`. Followed by

>   `dvc checkout` to update DVC-tracked files and directories in your workspace.

Author: This is basically part of the "Experiments" section. Let's merge them?

Reviewer: Done in new update.

Author: Thanks. It's too long now 😆 but anyway, let's address in the separate PR you're going to create. Resolving

- Switching between datasets

  You can quickly switch between a large dataset and a small subset without
  modifying source code. To achieve this yoe need to change dependencies of
  relevant stage either by using `dvc run` with `-f` option or by manually

>   editing the stage in `dvc.yaml` file.

Author: Where did this come from? Not seeing how it's a best practice. Kind of just a tip but this is covered in multiple other docs.

Author: Oh I see it comes from "How do you quickly switch between a large dataset and a small subset without
  modifying source code?" - that was part of the old **Navigating through experiments** question group, which again, is really part of the **Experiments** section.

Let's think things through please, not just keep content for the sake of keeping it. Thanks

Reviewer: Moved this to `tip-and-tricks` doc as its just a simple trick.

- Manually editing dvc.yaml or .dvc files

  It's safe to edit `dvc.yaml` and `.dvc` files. You can manually change all the
  fields present in these files. However, please keep in mind to not change the
  `md5` or `checksum` fields in `.dvc` files as they contain hash values which
  DVC uses to track the file or directory.

- Never store credentials in project config

  Do not store any user credentials in project config file. This file can be
  found by default in `.dvc/config`. Use `--local`, `--global`, or `--system`
  command options with `dvc config` for storing sensitive, or user-specific
  settings:

  ```dvc
  $ dvc config --system remote.username [password]
  ```

- Tracking <abbr>outputs</abbr> by Git

  If `outs` are small files in size and you want to track them with Git then you
  can use `--outs-no-cache` option to define outputs while creating or modifying
  a stage. DVC will not track will not track outputs in this case:

  ```dvc
  $ dvc run -n train -d src/train.py -d data/features \
            ---outs-no-cache model.p \
            python src/train.py data/features model.pkl
  ```

---

## Questions on...

### Source code and data versioning

> - How do you avoid discrepancies between

Author: Looks like everything under here is repeated now. It's fine to remove it if it has been rewritten in a how-to format.

# Best Practices for DVC Projects

Data scientists, engineers, or managers may already know or can easily find
answers to some of these questions. However, the variety of answers and
approaches makes data science collaboration a nightmare. **A systematic approach
is required.**

## Questions on...

### Source code and data versioning

> - How do you avoid discrepancies between

Author: @imhardikj this is a WIP begining to address #72 (listed in #899). May be a good first one to try and finish along with this PR.

# Best Practices for DVC Projects

Data scientists, engineers, or managers may already know or can easily find
answers to some of these questions. However, the variety of answers and
approaches makes data science collaboration a nightmare. **A systematic approach

> is required.**

Author: This text doesn't seem to match the title.

> I have updated the doc and now its similar to QnA. The answers reflect what is the best way to achieve the solution to the question. This can be a useful doc for users. Thoughts on the format?

The contents are improving, thanks. But a QnA is not a guide. We had mentioned the idea of leaving a section for FAQs at the bottom of Best Practices doc, but not replace it completely.

I don't think this doc should be in QnA form. It's a how-to guide. See https://documentation.divio.com/how-to-guides/#how-to and reshape the doc that way.

If there are any topics left that can only or much better be expressed as question-and-answer, then let's create yet another separate page for that (hidden from nav for now).

Thanks

Reviewer: QnA part was just for experiment. But I get the gist of how the best practices doc should look like (to the point solutions). I have  changed the starting para and also described each problem as a bullet point with a heading(heading could be improved more). Have added code blocks to some points. If this format looks fine then I can proceed to add more code blocks. Thanks.
Ps: the question in last part are same as those at start of this PR

Author: OK I reviewed the current format and left some comments.

  [revisions](https://git-scm.com/docs/revisions) of source code and versions of
  data files, when the data cannot fit into a traditional repository?

### Experiments

- How do you document progress of training different types of models on your

>   data files in the same project?

Reviewer: If am not wrong this material was previously named as Collaboration issues. I think this isn't required here. But if we want to keep this it can be under be a document under User guide with something like 'Why DVC?' but to not required to be honest.

>   data files in the same project?

Reviewer: I have updated this doc with some checkboxes from #72 but haven't deleted this material yet.

>   data files in the same project?

Author: I don't think it's from **Collaboration Issues** but it doesn't matter. Maybe leave it for now as a FAQ section at the bottom of this guide, but several of these questions could be turned into practices for sure e.g. "How do you document progress of training different types of models?"

>   data files in the same project?

Author: But my question is what is your plan for this doc, after having read these questions, and issue #72? BTW you may want to also search "practice", "best practice", "recommended way", and similar phrases in the Discord channel to look for other common questions that users have.

>   data files in the same project?

Reviewer: Yeah I thought this doc should address checkboxes in #72 . But now realised they are separate issues and best practices is a different doc. I have updated the doc and now its similar to QnA. The answers reflect what is the best way to achieve the solution to the question. Thoughts on the format?

>   data files in the same project?

Author: It's an interesting format but there's some confusion. Maybe I shouldn't have combined both issues in this PR, sorry. And yes, they are separate issues (start How-to section #899 and write a Best Practices guide #72) but closely related because you can't start a How-to section without having a few how-to guides, and I think The Best Practices guide belongs in this section (along with the other guide I already moved into /how-to when creating this this PR). See?

>   data files in the same project?

Author: > best practices is a different doc

Summarizing: How-to is not a doc., but a section. Best Practices is the first new doc in that section.

>   data files in the same project?

Author: But my questions on the new Best Practices doc remain, please address them directly: what is your plan for this doc? Which best practices have you researched (or need help with)? Thanks

>   data files in the same project?

Reviewer: I think Best Practices doc looks good as first doc in How-to section. I have added a point on how to track outs by Git instead of DVC(--out-no-cache in `run`). I think another one can be best way to import DVC tracked files from other DVC repos(import stage). Apart from these i'm searching for more.

>   data files in the same project?

Author: Asking again:
- What is your plan for this doc? I'm still not seeing a clear direction.
- Which best practices have you researched (or need help with)?

>   data files in the same project?

Reviewer: in my view this doc will be better under user-guide directly. It should be single doc. There will be some best practices (like **Managing and sharing large data files**) which already have examples in other docs so we can explain the best practice to follow and then link it to the already existing example. single doc should be good and it can be more generalized. If a best practice is big enough then it can be a separate doc directly under how-to section.

>   data files in the same project?

Author: OK @imhardikj, in that case please extract the best practices guide into a separate draft PR (use "WIP" in the title for now). That PR should aim to completely close #72.

> If a best practice is big enough then it can be a separate doc directly under how-to section.

That didn't make much sense to me but OK, not something we have to decide right now.

>   data files in the same project?

Reviewer: Have created the new PR #1748. Answered https://github.com/iterative/dvc.org/pull/1705#issuecomment-683352601 here only.
This PR can be merged after moving existing how-to's to this section. Apart from `Updating Tracked Files`, `running on windows` can also be moved here. WDYT
But I think it would be better to add more docs in this sub section  based on issues in #899. This way this PR will be substantial and have much more content. Thanks.

>   data files in the same project?

Author: > This PR can be merged after moving existing how-to's to this section

Oh no, that's just the beginng haha. We need to add some new content 🙂 pick something from #899

> running on windows can also be moved here. WDYT

Nah, that content is more like tips. Not how-to content (goal oriented). Again, please review this so it's extra clear what a how to is: https://documentation.divio.com/how-to-guides/#how-to

You're right that one doc as well as others like **Related Technologies** may not really belong in _User Guide_ though, but that's a completely separate problem, let's not further complicate this task for now 🙂

>   data files in the same project?

Author: > Answered https://github.com/iterative/dvc.org/pull/1705#issuecomment-683352601 here only.

What was the answer? 🙂

>   data files in the same project?

Reviewer: > What was the answer

@jorgeorpinel  I meant that it would be better to add more docs based on issues in #899 in this PR. That's the plan for this PR.

I have added a how-to doc based on #625 .
